PT-1920 | Add critical path Playwright tests for creating an event

karisal-anders opened the pull request:
## Description :sparkles:

NOTE:
 - [README](https://github.com/City-of-Helsinki/palvelutarjotin-admin/blob/98a68a4eedc7c79c059eb0afa4f965176a107d46/src/playwright/README.md#setup) has info on the test setup

Add Playwright critical path tests for event creation.

Also:
- Remove use of broken & unused
  NEXT_PUBLIC_CMS_TERMS_OF_SERVICE_SLUG_(FI|EN|SV) environment variables
  - This was done because Playwright couldn't run successfully with the
    import.meta.env.NEXT_PUBLIC_CMS_TERMS_OF_SERVICE_SLUG_* for some
    reason
- Fix / normalize some Swedish translations
- Add data-testid-values to EventForm/CreateOccurrencePage/LocationForm
  parts for ease-of-use in Playwright tests
- Regenerate generated GraphQL types and fix their gql imports to work
  with Playwright
- Add mocks for Playwright tests, mostly copy-pasted from actual network
  traffic and manually modified for test purposes
- Remove running Playwright tests from Github actions, as now the tests
  require a secret and that can't be easily provided in that environment
  currently i.e. Playwright tests need to be moved to use City of
  Helsinki's Azure DevOps environment instead

## Issues :bug:

### Closes :no_good_woman:

### Related :handshake:

[PT-1920](https://helsinkisolutionoffice.atlassian.net/browse/PT-1920)

## Testing :alembic:

### Automated tests :gear:️

Needs to be set up in Azure DevOps CI/CD pipelines, not yet running there.
Please see **[README](https://github.com/City-of-Helsinki/palvelutarjotin-admin/pull/416/files#diff-5aa662a171a01d96ded507d2f883d20fc68e77a888274297686da50284fa8105)** for the **required environment variables**.

### Manual testing :construction_worker_man:

Ran the tests locally using local backend & test environment CMS and test environment Linked Events.
The tests passed using parallel run i.e. `yarn test:browser:ci`:
```
$ yarn test:browser:ci
yarn run v1.22.22
$ playwright test

Running 38 tests using 8 workers
…\tests\eventCreationStep2.spec.ts:10:3 › create event's 2nd page opens in correct state with data saved on 1st page using language sv
Setting up auth service intercept for OIDC requests as user Test-admin
...
...
...
Mocking GraphQL operation "menu" response
Generated test JWT for user Test-admin
Generated test JWT for user Test-admin
  38 passed (39.6s)

To open last HTML report run:

  yarn playwright show-report

Done in 40.77s.
```

## Screenshots :camera_flash:

## Additional notes :spiral_notepad:


[PT-1920]: https://helsinkisolutionoffice.atlassian.net/browse/PT-1920?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela commented:
This one test (in every language) fails because it checks the LinkedEvents internal id which can vary in different environments:

# Instructions

- Following Playwright test failed.
- Explain why, be concise, respect Playwright best practices.
- Provide a snippet of code with the fix, if possible.
# Test info

- Name: a new event's occurrences (page 2/3) can be added using language fi
- Location: /Users/nikojmakela/Code/ratkaisutoimisto/palvelutarjotin-admin/src/playwright/tests/eventCreationStep2.spec.ts:168:3

# Error details

```
Error: expect(received).toEqual(expected) // deep equality

- Expected  - 2
+ Received  + 2

@@ -1,6 +1,6 @@
- ObjectContaining {
+ Object {
    "event": Object {
      "audience": Array [
        Object {
          "internalId": "https://linkedevents.api.test.hel.ninja/v1/keyword/kultus:40/",
        },
@@ -34,11 +34,11 @@
        Object {
          "internalId": "https://linkedevents.api.test.hel.ninja/v1/keyword/yso:p18434/",
        },
      ],
      "location": Object {
-       "internalId": "https://linkedevents.api.test.hel.ninja/v1/place/tprek:7254/",
+       "internalId": "http://localhost:8080/v1/place/tprek:7254/",
      },
      "name": Object {
        "en": "",
        "fi": "Testitapahtuman nimi",
        "sv": "",
    at /Users/nikojmakela/Code/ratkaisutoimisto/palvelutarjotin-admin/src/playwright/tests/eventCreationStep2.spec.ts:335:32
```

# Page snapshot

```yaml
- banner:
  - link "Siirry sisältöön":
    - /url: "#main-content"
  - link "Helsingin kaupunki":
    - /url: "#"
    - img "Helsinki"
  - link "Kultus Admin":
    - /url: "#"
  - button "Suomi"
  - button "Svenska"
  - button "English"
  - button "Avaa käyttäjän valikko": Test-admin
  - navigation:
    - list:
      - listitem:
        - link "Mikä kultus?":
          - /url: /fi/cms-page/mika-kultus-2/
        - button "Avaa alasvetovalikko."
      - listitem:
        - link "Kulttuurikasvatus":
          - /url: /fi/cms-page/kulttuurikasvatus/
        - button "Avaa alasvetovalikko."
      - listitem:
        - link "Oppimateriaalit":
          - /url: /fi/cms-page/oppimateriaalit-2/
        - button "Avaa alasvetovalikko."
- img
- contentinfo:
  - heading "Kultus Admin" [level=2]
  - img "Helsingin kaupunki"
  - text: © Copyright 2025 • Kaikki oikeudet pidätetään
  - link "Kouluille ja päiväkodeille":
    - /url: https://kultus-ui.test.hel.ninja/fi
  - link "Rekisteriseloste":
    - /url: https://hkih.production.geniem.io/uploads/sites/5/2022/11/739f5edc-rekisteriseloste-kultus.fi_.pdf
  - link "Takaisin alkuun"
- region "Notifications Alt+T":
  - alert:
    - img
    - text: Tapahtuma-aika tallennettu
    - button "close"
    - progressbar "notification timer"
  - alert:
    - img
    - text: Tiedot tallennettu
    - button "close"
    - progressbar "notification timer"
- status
```

# Test source

```ts
  235 |     await p2.fillTextBox(p2.necessaryVisitsSpinButton, '1');
  236 |
  237 |     await p2.autoAcceptRegistrationsCheckbox.check();
  238 |
  239 |     await expect(p2.autoAcceptanceMessageTextBox).toBeVisible();
  240 |     await expect(p2.autoAcceptanceMessageTextBox).toBeEnabled();
  241 |     await expect(p2.autoAcceptanceMessageTextBox).toBeEmpty();
  242 |
  243 |     p2.fillTextBox(
  244 |       p2.autoAcceptanceMessageTextBox,
  245 |       'Peruutukset test@example.org viim. 2pv ennen tapahtuma-aikaa, kiitos!'
  246 |     );
  247 |
  248 |     //-----------------------------------------------------------------
  249 |     // Fill "Occurrences" section data
  250 |     //-----------------------------------------------------------------
  251 |
  252 |     // Occurrence location should be pre-filled with the selected default venue
  253 |     await p2.hasSelectedOccurrenceLocation(p2.t('annantalo'));
  254 |
  255 |     // A date that won't run out during this application's lifetime
  256 |     await p2.fillTextBox(p2.occurrenceDateTextBox, '21.6.2200');
  257 |
  258 |     await p2.fillTextBox(p2.occurrenceStartHoursTextBox, '8');
  259 |     await p2.fillTextBox(p2.occurrenceStartMinutesTextBox, '0');
  260 |
  261 |     await p2.fillTextBox(p2.occurrenceEndHoursTextBox, '09');
  262 |     await p2.fillTextBox(p2.occurrenceEndMinutesTextBox, '00');
  263 |
  264 |     // Select occurrence languages
  265 |     await p2.openDropdown(p2.occurrenceLanguageDropdownButton);
  266 |     const langListBox = p2.occurrenceLanguageListBox;
  267 |     for (const language of p2.possibleOccurrenceLanguages) {
  268 |       await p2.hasListBoxOption(langListBox, language);
  269 |     }
  270 |     await p2.checkListBoxOption(langListBox, p2.t('finnishLanguage'));
  271 |     await p2.checkListBoxOption(langListBox, p2.t('swedishLanguage'));
  272 |     await p2.checkListBoxOption(langListBox, p2.t('englishLanguage'));
  273 |     await p2.closeDropdown(p2.occurrenceLanguageDropdownButton);
  274 |
  275 |     await p2.fillTextBox(p2.totalSeatsSpinButton, '80');
  276 |     await p2.fillTextBox(p2.minNumPersonsSpinButton, '2');
  277 |     await p2.fillTextBox(p2.maxNumPersonsSpinButton, '10');
  278 |
  279 |     //-----------------------------------------------------------------
  280 |     // Add the occurrence to the event
  281 |     //-----------------------------------------------------------------
  282 |     const addOccurrenceRequestPromise =
  283 |       p2.waitForGraphQLRequest('AddOccurrence');
  284 |
  285 |     await expect(p2.occurrencesTable).toBeHidden();
  286 |     await p2.addNewOccurrenceButton.click();
  287 |     await p2.isReady();
  288 |     await expect(p2.occurrencesTable).toBeVisible();
  289 |     await expect(p2.occurrencesTableRows).toHaveCount(2); // Header & occurrence row
  290 |
  291 |     //-----------------------------------------------------------------
  292 |     // Test that AddOccurrence request is sent using the filled data
  293 |     //-----------------------------------------------------------------
  294 |     const addOccurrenceRequest = await addOccurrenceRequestPromise;
  295 |     const addOccurrenceVariables =
  296 |       addOccurrenceRequest.postDataJSON().variables;
  297 |
  298 |     expect(addOccurrenceVariables).toEqual(
  299 |       expect.objectContaining({
  300 |         input: {
  301 |           startTime: '2200-06-21T05:00:00.000Z',
  302 |           endTime: '2200-06-21T06:00:00.000Z',
  303 |           languages: [{ id: 'fi' }, { id: 'sv' }, { id: 'en' }],
  304 |           pEventId: 'UGFsdmVsdXRhcmpvdGluRXZlbnROb2RlOjE1NjY=',
  305 |           placeId: 'tprek:7254',
  306 |           amountOfSeats: 80,
  307 |           minGroupSize: 2,
  308 |           maxGroupSize: 10,
  309 |           seatType: 'CHILDREN_COUNT',
  310 |         },
  311 |       })
  312 |     );
  313 |
  314 |     //-----------------------------------------------------------------
  315 |     // Go to the next step
  316 |     //-----------------------------------------------------------------
  317 |
  318 |     // Start waiting for the expected upcoming mutations
  319 |     const editEventRequestPromise = p2.waitForGraphQLRequest('EditEvent');
  320 |     const editVenueRequestPromise = p2.waitForGraphQLRequest('EditVenue');
  321 |
  322 |     // Override the already mocked event query response to return the new data at page 2 save
  323 |     // i.e. including the added occurrence and the edited venue:
  324 |     await p2.mockGraphQL({
  325 |       operationName: 'Event',
  326 |       responseData: MOCK_EVENT_QUERY_RESPONSE_AFTER_PAGE_2_SAVE,
  327 |     });
  328 |
  329 |     await p2.goToPublishingButton.click();
  330 |
  331 |     // Test that the EditEvent request was sent with the correct data
  332 |     const editEventRequest = await editEventRequestPromise;
  333 |     const editEventVariables = editEventRequest.postDataJSON().variables;
  334 |
> 335 |     expect(editEventVariables).toEqual(
      |                                ^ Error: expect(received).toEqual(expected) // deep equality
  336 |       expect.objectContaining({
  337 |         event: {
  338 |           id: 'kultus:agls75a3jm',
  339 |           name: {
  340 |             en: '',
  341 |             fi: 'Testitapahtuman nimi',
  342 |             sv: '',
  343 |           },
  344 |           startTime: '2025-06-02T09:23:35.687000Z',
  345 |           endTime: null,
  346 |           offers: [
  347 |             {
  348 |               isFree: false,
  349 |               description: {
  350 |                 en: '',
  351 |                 fi: 'Vain korttimaksu',
  352 |                 sv: '',
  353 |               },
  354 |               price: {
  355 |                 en: '10',
  356 |                 fi: '10',
  357 |                 sv: '10',
  358 |               },
  359 |               infoUrl: null,
  360 |             },
  361 |           ],
  362 |           shortDescription: {
  363 |             en: '',
  364 |             fi: 'Lyhyt kuvaus',
  365 |             sv: '',
  366 |           },
  367 |           description: {
  368 |             en: '',
  369 |             fi: '<p>Pitkä kuvaus</p>\n',
  370 |             sv: '',
  371 |           },
  372 |           images: [],
  373 |           infoUrl: {
  374 |             en: '',
  375 |             fi: 'https://example.org/testi/',
  376 |             sv: '',
  377 |           },
  378 |           audience: [
  379 |             {
  380 |               internalId:
  381 |                 'https://linkedevents.api.test.hel.ninja/v1/keyword/kultus:40/',
  382 |             },
  383 |             {
  384 |               internalId:
  385 |                 'https://linkedevents.api.test.hel.ninja/v1/keyword/kultus:58/',
  386 |             },
  387 |           ],
  388 |           inLanguage: [],
  389 |           keywords: [
  390 |             {
  391 |               internalId:
  392 |                 'https://linkedevents.api.test.hel.ninja/v1/keyword/kultus:17/',
  393 |             },
  394 |             {
  395 |               internalId:
  396 |                 'https://linkedevents.api.test.hel.ninja/v1/keyword/kultus:5/',
  397 |             },
  398 |             {
  399 |               internalId:
  400 |                 'https://linkedevents.api.test.hel.ninja/v1/keyword/yso:p18434/',
  401 |             },
  402 |           ],
  403 |           location: {
  404 |             internalId:
  405 |               'https://linkedevents.api.test.hel.ninja/v1/place/tprek:7254/',
  406 |           },
  407 |           draft: true,
  408 |           organisationId: 'T3JnYW5pc2F0aW9uTm9kZTox',
  409 |           pEvent: {
  410 |             contactEmail: 'admin@example.com',
  411 |             contactPersonId:
  412 |               'UGVyc29uTm9kZTpjZjNkN2NlYS0xNGQxLTQzNmQtOTNmZS0zOTI4Mzc2ZTQ2YTE=',
  413 |             contactPhoneNumber: '+358-50-123456789',
  414 |             mandatoryAdditionalInformation: false,
  415 |             enrolmentStart: '2199-12-31T14:30:00.000Z',
  416 |             enrolmentEndDays: 2,
  417 |             neededOccurrences: 1,
  418 |             autoAcceptance: true,
  419 |             translations: [
  420 |               {
  421 |                 languageCode: 'FI',
  422 |                 autoAcceptanceMessage:
  423 |                   'Peruutukset test@example.org viim. 2pv ennen tapahtuma-aikaa, kiitos!',
  424 |               },
  425 |             ],
  426 |             externalEnrolmentUrl: null,
  427 |           },
  428 |         },
  429 |       })
  430 |     );
  431 |
  432 |     // Test that the EditVenue request was sent with the correct data
  433 |     const editVenueRequest = await editVenueRequestPromise;
  434 |     const editVenueVariables = editVenueRequest.postDataJSON().variables;
  435 |
```

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders commented on src/playwright/mocks.ts:
```diff
@@ -0,0 +1,1812 @@
+  PlacesQuery,
+  PublishSingleEventMutation,
+  VenueQuery,
+} from '../generated/graphql';
```
Split the `mocks.ts` file into ~10 files under mocks directory. So, taken into account.

karisal-anders commented on fix-codegen-gql-import.js:
```diff
@@ -0,0 +1,31 @@
+    // eslint-disable-next-line no-console
+    console.info(`Fixed gql imports in ${file}`);
+  }
+});
```
Fixed properly with codegen.yml configuration in [force push](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/7329e0a5ee13eaf38eca913b42431f8779388832..a8f98a973b4d4e6f0ae2b1a67581dd07e82d46a8).

karisal-anders commented on src/generated/graphql.tsx:
```diff
@@ -1,4 +1,4 @@
-import { gql } from '@apollo/client';
+import { gql } from 'graphql-tag';
```
Fixed properly with codegen.yml configuration in [force push](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/7329e0a5ee13eaf38eca913b42431f8779388832..a8f98a973b4d4e6f0ae2b1a67581dd07e82d46a8) i.e. importing gql from graphql-tag here, but it is used a lot in this codebase already so it should be ok.

karisal-anders commented:
This is now ready for code review & possible merging. I added more to the README also so now the necessary environment variables should be documented and they should help with setting up these tests to Azure DevOps CI/CD pipelines (FYI @terovirtanen) as well as locally.

Please @nikomakela could you have another look now after all the changes? Thanks!

@City-of-Helsinki/ratkaisutoimiston-frontend As palvelutarjotin was officially received by Ratkaisutoimisto on Monday, please could you code review this PR also? Thank you!

nikomakela commented:
Quite much changes overall, but looked good to me.

karisal-anders commented:
Cleanup [force push](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/fa701b8b9483212c61a3a29feeefc0d6dbe02032..98a68a4eedc7c79c059eb0afa4f965176a107d46):
- Anonymized mock phone numbers
- Clarified comment in myProfile.spec.ts

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders commented:
[Force pushed](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/98a68a4eedc7c79c059eb0afa4f965176a107d46..2fc10e538e2c96eeb0db23b1b4d3fb48e0a7a19a) enabling Playwright web server to always reuse existing server.

This was done to help set up the Playwright tests in Azure DevOps CI/CD pipelines.

Related Slack discussion on City of Helsinki's Slack:
- https://helsinkicity.slack.com/archives/C04DLNKDF40/p1749816210582209?thread_ts=1749798612.400199&cid=C04DLNKDF40

terovirtanen commented:
**KULTUS-ADMIN-UI branch is deployed to platta: https://kultus-admin-ui-pr416.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders commented:
[Force pushed](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/2fc10e538e2c96eeb0db23b1b4d3fb48e0a7a19a..da18552912803da16abdc2f8f86ff1fe37970aa3):
- Moved `playwright` and `@playwright/test` to development dependencies
- Added `test:browser:install` command for installing browsers for Playwright (Same syntax format as `test:browser:ci` etc)
- Updated Playwright tests' README related to use of `test:browser:install` (the browser installing was there, but now it uses this command)

karisal-anders commented:
[Force pushed](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/21c2d1712228d81408e0d2b423e0e3aa033bd518..2a118bc8bd07b88ef68a16dccba75514a0c4cca8):
- Fix to exclude Playwright 1.53.0 from `playwright` and `@playwright/test` packages' versions because of [regression](https://github.com/microsoft/playwright/issues/36307) that caused some of the tests to fail
  - Added comment about why this was done to the comment section in package.json (i.e. under the `"//"` key) for clarity

karisal-anders commented:
[Force pushed](https://github.com/City-of-Helsinki/palvelutarjotin-admin/compare/2a118bc8bd07b88ef68a16dccba75514a0c4cca8..b21163704b7da2c7893193c46b111f26ff745743):
- Configure Playwright tests
  - Set [outputDir](https://playwright.dev/docs/api/class-testconfig#test-config-output-dir) to './report/test-results'
  - Set `junit` and `html` reports
    - junit → report/playwright-junit-results.xml
    - html -> report/html
  - Enable screenshots `only-on-failure` and video recording `on-first-retry`
